### ISSUE TITLE
Feature request: add API to customise front-end HTML rendering of rich text features

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#4223]
### Issue Summary

New feature: add an additional "rich text features" API to give developers full control over how the front-end HTML is created, for custom and also built-in features.

At the moment, unless the feature is a link with a linktype or an embed (images and actual embeds), we can only define how it is stored as HTML in the DB. This means that the feature's "storage format" must be the same as its "front-end format". This is too limiting – for example, it should be possible to add a class attribute to the feature's front-end output without having it stored in the DB. It should be possible to store a `strikethrough` format as `<s>` tags but render them on the front-end as a more semantic `<del>`, should it make more sense for a given site.

Here is what the current APIs allows you to do:

```python
features.register_converter_rule(
        'contentstate', 'blockquote', {
            [...]
            'to_database_format': {
                'block_map': {'blockquote': {
                    'element': 'blockquote',
                    # This is in to_database_format, so the presentational attributes will be stored.
                    'props': {
                        'class': 'richtext__blockquote',
                    }
                }}
            }
        })
```

Ideally, there should be a separate `to_frontend_html` (or equivalent) to control the HTML on the front-end only. Note: it might not make sense to put this in `register_converter_rule`, since these are specific to a given converter (editorhtml for Hallo, contentstate for Draftail), whereas the new feature wouldn't be.

### Use cases

Here are the use cases we had when building [WagtailDraftail](https://github.com/springload/wagtaildraftail), beyond custom links and embeds:

```py
DRAFT_EXPORTER_BLOCK_MAP = dict(BLOCK_MAP, **{
    BLOCK_TYPES.UNORDERED_LIST_ITEM: {
        'element': 'li',
        'wrapper': 'ul',
        # Adding a class to rich text elements to make styling easier.
        'wrapper_props': {'class': 'list-styled'},
    },
    INTRO_BLOCK_ID: {
        'element': 'p',
        # Having a rich text format that's based on the same tag as other formats, + some classes.
        'props': {'class': 'intro'},
    },
})
```

### Context

This is the last missing piece to bring Wagtail's Draftail integration to parity with [WagtailDraftail](https://github.com/springload/wagtaildraftail) (and make that plugin useless, _yay!_).

It would also:

- Support what is requested in #3257 (and thus #1167).
- Help with mitigating #1214.
- Potentially replace the undocumented `register_rich_text_embed_handler` / `register_rich_text_link_handler` (#1094)

**Comment by SalahAdDin** [comment]
Actually in `Draftail` example `strikethrough` and `del` elements are different.

**Comment by thibaudcolas** [comment]
@SalahAdDin the editor doesn't use the semantic HTML tags, but once the content is saved it's all converted to semantic HTML (well, depending on how you configure it in `register_converter_rule`.

If you want to see how this works behind the scenes, http://draftjs-exporter.herokuapp.com/ is a good demo.

**Comment by SalahAdDin** [comment]
Ok, i understood, it's a powerful tool.

**Comment by thibaudcolas** [comment]
For future reference: https://github.com/wagtail/wagtail/pull/4527#issuecomment-389233975

**Comment by gasman** [comment]
Continuing the discussion from #4527 - I think the CSS selector approach we use for dbHTML -> contentState conversion would be a good fit for front-end HTML rewriting, since it allows us to generalise the rewrite rules to any element (instead of just two hard-coded cases for links and embeds) while still keeping the rules as a flat list (=> only one hook required to extend them, rather than it being a big messy decision tree that could potentially be extended at any point).

The problem is that our CSS selector code in `wagtail.admin.rich_text.converters.html_ruleset` requires us to run the text through an HTML parser first, and in my (very brief) tests that was something like 5-10x slower than the current regexp-based approach. It may not be a big deal, but from past experience I'm wary of making any assumptions like "surely rich text rendering isn't a performance bottleneck on real-world Wagtail sites, right?" :-)

I think this can be solved with a hybrid approach, where we initially use a regexp to match all the tags we're interested in, and in the (hopefully rare and small) cases where the rewriting requires us to find the corresponding closing tag, switch to proper tag-by-tag HTML parsing. Unfortunately I think this will require a bit of reinventing the wheel (since we can't hack the behaviour of Python's `html.parser` to say "stop parsing when you encounter the closing tag that matches the one you started with, and hand control back to us") - I reckon it's doable in a couple of days' development, though.

**Comment by thibaudcolas** [comment]
I’ve thought of a few more use cases for the above based on recent discussions:

- #4665 – if what’s highlighted here was implemented, switching from `b` to `strong` would just be a matter of changing the HTML rendering’s configuration, rather than having to migrate all of the site’s content.
- Comments/annotations UIs, and generally anything that should be displayed conditionally – it might make sense for some of the content’s (meta)data to only be visible in the CMS admin, but currently there is no way to have this type of metadata stored with the content and then not displayed on the front-end.

**Comment by loicteixeira** [comment]
So there are 2 sides to this problem, the API side of things (how it is exposed to users) and the inner working of it (hot it is applied.)

Regarding the former, I agree with adding a `to_frontend_html` key in the `register_converter_rule` hook options. It feels like the best place for it.

As for the implementation, there has been many discussions (both online and offline), so to summarise, here are the different solutions which were mentionned:

1. Save the editor *JSON content state* instead of *DB HTML*
    - The content state can be rendered with any configuration, no transformation needed
    - However the transformation would happen when rendering the templates
    - It would require migrating the data of all rich-text fields and blocks
    - We should note that this option was discarded when WagtailDraftail was merged into core (despite being how it worked in the module) for backward compatibility reasons and to avoid vendor lock-in
2. Save both the current *DB HTML* representation and the editor content-state as JSON
    - Can it be 2 DB fields or does it have to be 1 (assuming it's still just 1 model field)?
        - The former is straightforward from a migration POV but more complexe on the Python side.
        - Might or might not need migrations (if the field can handle missing data)
    - Bigger database
    - Risk of fields not being in sync?
    - The transformation would still happen when rendering the template
3. Save both the current *DB HTML* representation and the *front-end HTML*
    - Technically similar to option 2
    - The transformation would have happened already
    - However updating the code will not update the rendering, a page would need to be saved again
4. Add more regexp based rewriters options, similarly to the links & embeds ones.
    - Most of it is already there
    - No migration needed
5. Add [CSS selector based rewriters](https://github.com/wagtail/wagtail/issues/4223#issuecomment-389483114), as discussed in the comment above.
    - No migration needed
    - Quite slow with first tests
6. Recreate the *JSON content state* from the *DB HTML* (like the editor is doing when reading from the DB), then convert to *front-end HTML* with a specific configuration.
    - No migration needed
    - Can be slow
    - But it's the same mechanism as when the editor saves to the database, only with a different configuration. So that's an already well understood concept. Users would set up the `to_frontend_html` similarly to `to_database_html` (but it could also be confusing?)

**Comment by loicteixeira** [comment]
Sorry about the repeated reference notification 👆 😰

So loicteixeira/wagtail@428f1081b53af083dec0bd3d99700039a5ce69df implements option 6 and here is the result of a quick benchmark:

| Page                  | Notes                                                                                   | Measure | Original Time in ms | Patched Time in ms | Diff in % |
|-----------------------|-----------------------------------------------------------------------------------------|---------|---------------------|--------------------|-----------|
| /about/               | Original Bakerydemo page. 1 rich-text block (body) 1 rich-text field (footer)           | MIN     | 47.9                | 47.12              | - 1.6     |
| "                     | "                                                                                       | MAX     | 142.59              | 175.66             | +23.19    |
| "                     | "                                                                                       | MEAN    | 66.9296             | 71.9928            | + 7.56    |
| /blog/bread-circuses/ | Original Bakerydemo page. 1 rich-text block (body) 1 rich-text field (footer)           | MIN     | 84.94               | 86.97              | + 2.38    |
| "                     | "                                                                                       | MAX     | 190.38              | 307.29             | +61.40    |
| "                     | "                                                                                       | MEAN    | 106.972             | 136.044            | +27.17    |
| /new-page/            | Manually created `Standard Page`. 10 rich-text blocks (body) 1 rich-text field (footer) | MIN     | 61.67               | 67.13              | + 8.85    |
| "                     | "                                                                                       | MAX     | 267.2               | 175.4              | -34.36    |
| "                     | "                                                                                       | MEAN    | 91.086              | 102.6508           | +12.69    |

<details>
  <summary>Methodology</summary>

  With `debug_toolbar` and `template_profiler_panel` to the `bakerydemo` project, run the following script twice, once with Wagtail unchanged, and a second time with Wagtail patched.

  ```python
  import re
  from collections import defaultdict
  from statistics import mean
  from time import sleep
  
  import requests
  from bs4 import BeautifulSoup
  
  data = defaultdict(list)
  
  for endpoint in ("/about/", "/blog/bread-circuses/", "/new-page/"):
      for _ in range(25):
          response = requests.get(f"http://127.0.0.1:8000{endpoint}")
          soup = BeautifulSoup(response.text, "html.parser")
  
          panel_link_node = soup.find("a", {"class": "TemplateProfilerPanel"})
          panel_text = panel_link_node.find("small").text
          template_time = re.match(r'.*?(\d+\.\d+) ms$', panel_text).group(1)
  
          data[endpoint].append(float(template_time))
  
          sleep(0.25)
  
  for endpoint, datapoints in data.items():
      print(f"Endpoint: {endpoint}")
      print(f"  Min: {min(datapoints)}")
      print(f"  Max: {max(datapoints)}")
      print(f"  Mean: {mean(datapoints)}")
  ```
</details>

**Comment by easherma** [comment]
Would this have any relation to/implications for this issue as well? 
https://github.com/wagtail/wagtail/issues/2695
( we're running wagtail headless with graphql and expanding db html for RichTextBlocks remains a thorny issue)

**Comment by loicteixeira** [comment]
Not really. This issue is about how to have more control during the conversion from the *DB HTML* to the *front-end HTML*, but the issue you're facing with the API is that the conversion step isn't called.

That being said, if the resolution of #2695 was to call the conversion step, then enhancements provided by this issue would help you tweak the "final" HTML for your needs.

**Comment by thibaudcolas** [comment]
Thank you @loicteixeira, your comments above are invaluable! I’m surprised how little work this actually seems to be from your commit https://github.com/loicteixeira/wagtail/commit/428f1081b53af083dec0bd3d99700039a5ce69df, which I guess shows the benefit of using the existing converters for this rather than adding a new one.

I feel like it might not be enough to use `default_features` though – presumably that only contains things added as "default", but we’d need non-default features for each field too (`RichTextField(features=['italic', 'potato'])`). That’s one drawback of this approach compared to what we’d implement following the existing rewriters – with these you can apply the customisations regardless of what the fields contain, while `to_frontend_html` (and the other two conversion steps relied upon here) is tied to a rich text feature.

For your performance numbers – my benchmarks of the exporter show a contentstate -> html conversion time ranging from 0.1ms to 10ms per rich text field / block, depending on how much the field contains. 10ms is the maximum based on the content in my benchmark, based on a [field that contains 120 paragraphs](https://github.com/thibaudcolas/markov_draftjs#why) & heading levels + links. 0.1ms is the minimum, for a field that contains 3 text paragraphs & heading levels, about 500 characters.

**Comment by loicteixeira** [comment]
> I’m surprised how little work this actually seems to be from your commit [loicteixeira@428f108](https://github.com/loicteixeira/wagtail/commit/428f1081b53af083dec0bd3d99700039a5ce69df), which I guess shows the benefit of using the existing converters for this rather than adding a new one.

Indeed, although I am importing [`wagtail.admin` in a `wagtail.core` module](https://github.com/loicteixeira/wagtail/blob/428f1081b53af083dec0bd3d99700039a5ce69df/wagtail/core/rich_text/__init__.py#L28) so we might want to put a bit more work to avoid that (like moving the converter into the `wagtail.core` namespace since it's now use for more than just admin work).

> I feel like it might not be enough to use `default_features` though – presumably that only contains things added as "default", but we’d need non-default features for each field too.

Right. My intent was to include all the features (as my comment shows) but it was indeed short-sighted.

Looking at it, maybe we could use the converter rules instead, but not every feature has a converter rules (e.g. `b`) so we would need to both the default features and the converter rules. Something like that:

```diff
- all_existing_features = features.get_default_features()
+ default_features = features.get_default_features()
+ features_with_converter = list(features.converter_rules_by_converter['contentstate'].keys())
+ all_existing_features = list(set(default_features + features_with_converter))
```

But still, could we have tags which aren't a default feature but still don't have a converter rule?

**Comment by thibaudcolas** [comment]
My bad, you’re right, we can probably just use all the registered features (that have conversion rules) like you did.

For missing definitions – it looks like we have "fallbacks" defined for blocks and entities (but not inline styles), which simply log and discard the corresponding Draft.js block / entity: https://github.com/wagtail/wagtail/blob/982b1d60a4b3d5b4e841f2b39e2244bcb421b091/wagtail/admin/rich_text/converters/contentstate.py#L36-L45.

**Comment by gasman** [comment]
Converting dbHTML to contentstate and back to HTML seems like a very roundabout way to perform a few HTML element rewrites, and the "dbHTML to contentstate" part is exactly the CSS rule logic (i.e. your option 5) that I didn't think was performant enough to use on the front-end - so it seems odd that your benchmarking results are so much better than mine... I'll revisit this.

**Comment by gasman** [comment]
Haven't done any benchmarking yet, but I have a good feeling about this... :-)

https://gist.github.com/gasman/e640f6e12609ffe43718403203fd8d65

Rewrite rules defined by CSS selectors, which then get compiled to a regexp matching just the elements we intend to rewrite (so we avoid the overheads of full-scale HTML parsing of the entire document). Handles the existing link and embed rewrites in a generic way, as well as arbitrary processing of content inside a tag pair - including, hopefully, correct handling of nested tags.

**Comment by loicteixeira** [comment]
It looks promising 👍 

My only concern is introducing yet another system to interact with rich text. `to_database_format` is handled via `draftjs-exporter`, `from_database_format` is handled via `converters.html_to_contentstate` and now we would have some `html_rewriter` for the `to_frontend_format`.

That being said, it will probably perform better than a whole DB-HTML -> Content State -> Frontend-HTML solution. In addition, it seems that we already declare `from_database_format` elements matching with CSS selectors, so your solution would actually bring some unity on that front.

**Comment by gasman** [comment]
@loicteixeira This isn't really introducing a new system - it's replacing the existing limited functionality in https://github.com/wagtail/wagtail/blob/master/wagtail/core/rich_text/rewriters.py . The real advantage, though, is that it makes the front-end rewriter's API more consistent with the backend editor ones - for example, now that we're defining rules in terms of CSS selectors rather than hard-coded cases for `<embed embedtype="">` and `<a linktype="">`, FeatureRegistry no longer has to maintain a lookup of embedtypes / linktypes as a separate kind of thing to rich text features.

I'm hopeful that this will take us a step closer to being able to define rich text features in a more 'declarative' way whenever they follow a standard recipe (such as "a link that turns an object ID into a URL") - rather than having to define three or four converter functions using different mechanisms, it'll be a specification in the form of "this feature is represented by element X in the database, and element Y in frontend HTML; it appears as icon Z in the toolbar".

If the proliferation of parser implementations in the codebase is a concern, then (with a bit more development) I think we could potentially port the html_to_contentstate logic to this implementation. We could certainly port over the DB-HTML to hallo.js editor-HTML converter, if anyone still cares about that :-)

**Comment by kaedroho** [comment]
I'm working on an RFC to replace the HTML representation of rich text fields with a JSON based one that's based on Streamfield and a new ParagraphBlock type described in https://github.com/wagtail/rfcs/pull/46

This would mean rich text fields would have a rendering to HTML step, just like SrreamField. I think that could simplify the implementation of an API like this.

I also think that using JSON over HTML would also simplify the implementation of rich-text diffing, commenting, importers, and headless sites.

**Comment by kaedroho** [comment]
Adding a rendering step for rich text, like we already have for streamfield, would also make https://github.com/wagtail/wagtail/issues/6526 easy to solve

**Comment by thibaudcolas** [comment]
For people interested in this, PR extending our current rewriters approach with support for declaring arbitrary additional rewriters: #9292.
